src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -341,13 +489,77 @@ internal bool StartBufferingAndTryToReadInStreamErrorPropertyValue(out ODataErro
             }
         }
 
+        /// <summary>
+        /// Asynchronously puts the reader into the state where it buffers read nodes.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        internal async Task StartBufferingAsync()
+        {
+            Debug.Assert(!this.isBuffering, "Buffering is already turned on. Must not call StartBuffering again.");
+            this.AssertAsynchronous();
+
+            if (this.bufferedNodesHead == null)
+            {
+                // capture the current state of the reader as the first item in the buffer (if there are none)
+                object value = await this.asyncInnerReader.GetValueAsync()
+                    .ConfigureAwait(false);
+                this.bufferedNodesHead = new BufferedNode(this.asyncInnerReader.NodeType, value);

```

Comment:
I was curious about the relationship between `IJsonReader.NodeType` and `IJsonReader.Value` and whether accessing the value (or `GetValueAsync` could modify the node type). Based on my perusal of the code, it seems `NodeType` is only updated by `IJsonReader.Read()` (and ReadAsync). And `GetValueAsync()` does not actually end up calling `ReadAsync` (though it calls the inner text writer's `ReadAsync`). In short I don't think this is a problem, but I thought maybe you could confirm since you're more familiar with the code.

If it happens that accessing the value has a side effect on `NodeType`, then the order in which you access them matters (otherwise it doesn't). Because in this async method you're accessing the value by calling `GetValueAsync()` first, then the node type. But in the sync version, you're accessing `NodeType` first, then `Value`.

Reply:
@habbes `IJsonReader.Value` (and `IJsonReaderAsync.GetValueAsync()`) only read the value. They do not advance the reader. For this same reason, you can call them multiple times with the same result. All the same I have rewritten the logic to mirror the sync version.